### PR TITLE
Story at1 225

### PR DESCRIPTION
**Pull request by apurvapatkar**
Updated TMC Prototype to interface with CSP Master.
 - Monitor CSP Master attributes 
 - Invoke On command on CSP Master.

### CODE REVIEW

File: tmcprototype/CentralNode/CentralNode/CentralNode.py [hunk 1]
```diff
@@ -241,14 +258,27 @@ def init_device(self):
...
                 self.dev_logging(CONST.ERR_IN_CREATE_PROXY, int(tango.LogLevel.LOG_ERROR))
 
+        # Create device proxy for CSP Master Leaf Node
+        try:
+            self._csp_master_leaf_proxy = DeviceProxy(self.CspMasterLeafNodeFQDN[0])
+            self._csp_master_leaf_proxy.subscribe_event(CONST.EVT_SUBSR_CSP_MASTER_HEALTH,
+                                                        EventType.CHANGE_EVENT,
+                                                        self.healthStateCallback, stateless=True)
+        except Exception as except_occured:
+            print(CONST.ERR_SUBSR_CSP_MASTER_LEAF_HEALTH, self.CspMasterLeafNodeFQDN[0])
```
[the remark this answers is not in the document]
Comment (adityadangeska) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Check for other instances as well.

File: tmcprototype/CentralNode/CentralNode/CentralNode.py [hunk 2]
```diff
@@ -281,14 +312,14 @@ def delete_device(self):
...
         """ Returns the Telescope health state."""
-        if ((self._subarray1_health_state == 1) | (self._subarray2_health_state == 1)):
-            self._telescope_health = 1
-        elif ((self._subarray1_health_state == 2) & (self._subarray2_health_state == 2)):
-            self._telescope_health = 2
-        elif ((self._subarray1_health_state == 0) & (self._subarray2_health_state == 0)):
-            self._telescope_health = 0
-        else:
-            self._telescope_health = 3
+        # if ((self._subarray1_health_state == 1) | (self._subarray2_health_state == 1)):
```
Comment (adityadangeska) [inline tmcprototype/CentralNode/CentralNode/CentralNode.py]: Remove commented code or add TODO block if required in future.

File: tmcprototype/CentralNode/CentralNode/CentralNode.py [hunk 3]
```diff
@@ -372,6 +403,16 @@ def StandByTelescope(self):
                 print(CONST.STR_ERR_MSG, except_occured)
                 self._read_activity_message = CONST.STR_ERR_MSG + str(except_occured)
                 self.dev_logging(CONST.ERR_EXE_STANDBY_CMD, int(tango.LogLevel.LOG_ERROR))
+
+        try:
+            self._csp_master_leaf_proxy.command_inout(CONST.CMD_CSP_MASTER_LEAF_STANDBY, CONST.STR_CSP_CBF_DEV_NAME)
+        except Exception as except_occured:
+            print(CONST.ERR_EXE_STANDBY_CMD, self.CspMasterLeafNodeFQDN[0])
```
Comment (adityadangeska) [inline tmcprototype/CentralNode/CentralNode/CentralNode.py]: Consider consolidating multiple print statements in one if applicable. Ideally there is no need of print statements.

File: tmcprototype/CentralNode/CentralNode/CentralNode.py [hunk 4]
```diff
@@ -407,6 +459,7 @@ def StartUpTelescope(self):
     @DebugIt()
     def AssignResources(self, argin):
         # PROTECTED REGION ID(CentralNode.AssignResources) ENABLED START #
+
```
Comment (adityadangeska) [inline tmcprototype/CentralNode/CentralNode/CentralNode.py]: Why extra line break is added here?

File: tmcprototype/CentralNode/CentralNode/CentralNode.py [hunk 5]
```diff
@@ -262,6 +292,7 @@ def init_device(self):
                 print(CONST.STR_ERR_MSG, except_occured)
                 self._read_activity_message = CONST.STR_ERR_MSG + str(except_occured)
 
+
```
Comment (adityadangeska) [inline tmcprototype/CentralNode/CentralNode/CentralNode.py]: Why extra line break is added here?

File: tmcprototype/CspMasterLeafNode/setup.cfg [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# Define `python setup.py test`
+test=pytest
+
+[coverage:run]
+branch = True
+source = CspMasterLeafNode
+
+[tool:pytest]
+testpaths = test
+addopts = --cov=CspMasterLeafNode --json-report --json-report-file=htmlcov/report.json --cov-report term --cov-report html --cov-report xml
```
Comment (adityadangeska) [inline tmcprototype/CspMasterLeafNode/setup.cfg]: Line is very long. Check if it exceeds max length limit.

File: tmcprototype/CentralNode/CentralNode/CentralNode.py [hunk 7]
```diff
@@ -90,20 +93,30 @@ def subarrayHealthStateCallback(self, evt):
...
+                else:
+                    unknown_health_count = 1
+
                 for value in list(self.subarray_health_state_map.values()):
                     if value == CONST.ENUM_FAILED:
                         failed_health_count = failed_health_count + 1
                         break
                     elif value == CONST.ENUM_DEGRADED:
-                        self._telescope_health_state = 1
+                        #self._telescope_health_state = 1
```
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Done. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/dafa12fbac0e08db8ec3ec703afb99ba45db38c1

File: tmcprototype/CentralNode/CentralNode/CentralNode.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (apurvapatkar) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Done. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/dafa12fbac0e08db8ec3ec703afb99ba45db38c1

File: tmcprototype/CentralNode/CentralNode/CentralNode.py (hunk 5 above)
Comment (apurvapatkar) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Removed it. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/dafa12fbac0e08db8ec3ec703afb99ba45db38c1

File: tmcprototype/CentralNode/CentralNode/CentralNode.py (hunk 2 above)
Comment (apurvapatkar) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Done. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/dafa12fbac0e08db8ec3ec703afb99ba45db38c1

File: tmcprototype/CentralNode/CentralNode/CentralNode.py (hunk 3 above)
Comment (apurvapatkar) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Done wherever possible. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/dafa12fbac0e08db8ec3ec703afb99ba45db38c1

File: tmcprototype/CentralNode/CentralNode/CentralNode.py (hunk 4 above)
Comment (apurvapatkar) [reply tmcprototype/CentralNode/CentralNode/CentralNode.py]: Removed it. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/dafa12fbac0e08db8ec3ec703afb99ba45db38c1

File: tmcprototype/CspMasterLeafNode/setup.cfg (hunk 6 above)
Comment (apurvapatkar) [reply tmcprototype/CspMasterLeafNode/setup.cfg]: Done. The commit id is: https://github.com/ska-telescope/tmc-prototype/commit/530792d1807f498f367b47e6a1dada0e4a1cd50d

**Pull request merged**